blackjack.py

```diff
@@ -216,30 +121,125 @@ def bankrupt(self):
+            player.count.append(player.cards[0][0] + player.cards[1][0][1])
+        else:
+            player.count.append(player.cards[0][0] + player.cards[1][0])
+
+    def setdcount(self):
+        if isinstance(dealer.cards[0][0], list) == True and isinstance(dealer.cards[1][0], list) == True:
+            dealer.count = dealer.cards[0][0][1] + dealer.cards[1][0][1]
+        elif isinstance(dealer.cards[0][0], list) == True and isinstance(dealer.cards[1][0], list) == False:
+            dealer.count = dealer.cards[0][0][1] + dealer.cards[1][0]
+        elif isinstance(dealer.cards[0][0], list) == False and isinstance(dealer.cards[1][0], list) == True:
+            dealer.count = dealer.cards[0][0] + dealer.cards[1][0][1]
+        else:
+            self.count = dealer.cards[0][0] + dealer.cards[1][0]
+
+    def start(self):
+        print('Welcome to BlackJack')
+        name = input("What's your name? ")
+        print('Hello ' + name)
+        print('Drawing card...\n')
+        print(player.cards)
+        if game.counter == 0:
+            player.checkA()
+            self.setdcount()
+            self.setpcount()
+            game.counter += 1
+        player.check()
+        dealer.check() #how to stop these two from not restarting?
+
+    def restart(self):
+        while True:
+            again = input('\nWould you like to play again? (Y/N) ')
+            if again.lower() == 'y':
+                player.cards = [self.drawcards(), self.drawcards()]
+                game.setpcount()
+                dealer.cards = [self.drawcards(), self.drawcards()]
+                game.setdcount()
+                break
+            elif again.lower() == 'n':
+                sys.exit()
+            else:
+                print('Please enter Y or N')
+            print('')
+        player.bet()
+        self.counter = 0
+        return False
+
+    def stand(self):
+        dealer.hit()
+        dealer.check()
+        print('\nYour hand: ', player.cards)
+        print('Dealer\'s hand: ', dealer.cards)
+        if player.count[0] > dealer.count:
+            print('You won!')
+            player.bank += player.bid
+        elif player.count[0] == dealer.count:
+            print('You tied!')
+        else:
+            print('You lost!')
+            player.bank -= self.bid
+            self.bankrupt()
+
+    def spstand(self, num=0):
+        dealer.hit()
+        dealer.check()
+        print('\nYour hand: ', self.cards[num])
+        print('Dealer\'s hand: ', dealer.cards)
+        if player.count[num] > dealer.count:
+            print('You won!')
+            self.bank += self.bid
+        elif player.count[num] == dealer.count:
+            print('You tied!')
+        else:
+            print('You lost!')
+            self.bank -= self.bid
+            self.bankrupt()
+
 if __name__ == "__main__":       
-    game = GameManager()
     dealer = Dealer()
     player = Player()
+    game = GameManager()

```

Comment:
now `Dealer` and `Player` depends on `GameManager`. This is an awkward design, b/c now the agents in the game have to know about the whole game state and state transitions. Like I mentioned before, a better design would be to make `GameManager` depend on `Dealer` and `Player`. Then, `Dealer` only has to worry about dealing, and `Player` only has to worry about playing and not worry about having to restart the game.

Reply:
how is Dealer and Player depending on GameManager? Dealer only has to check & hit, Player only needs to check, hit, split, and bet; GameManager sets everything else up. Am I understanding something wrong? Dealer and Player doesn't have any restart methods -- only GameManager does

Reply:
`Dealer` and `Player` depend on `GameManager` b/c they call methods on `GameManager`. It should be that `GameManager` moves between different game states, calling appropriate methods on `Dealer` and `Player`. We should probably sit down and talk this over.